### PR TITLE
Rational numbers and a few additions to math and complex

### PR DESCRIPTION
**Pull request by def-**
Someone asked whether there is a rational numbers module in Nim, now there is.

The rest are additions that I commonly needed, so they may generally be useful.

### CODE REVIEW

**Comment by Varriount** [comment]
Please add a comment to the change log

**Comment by def-** [comment]
Not sure where, I only see the changelog up to 0.10.2 in web/news.txt

**Comment by reactormonk** [comment]
Why not as a Nimble package?

**Comment by def-** [comment]
For me it's in the same category as complex numbers, but I don't mind either way.

**Comment by dom96** [comment]
@Varriount @def- I usually look through the commit list anyway when a new release is being made so don't worry about that.

**Comment by ReneSac** [comment]
I think the module name should be "rationals", not "rational". Even though nim already has partial case sensitivity, plural seems to be the convention for module names. And I think it will be useful for specifying the clocks period/resolution on times.nim.

File: lib/pure/rational.nim [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+  result.den = 1
+
+proc toFloat*(x: Rational): float =
+  ## Convert a rational number `x` to a float.
+  float(x.num) / float(x.den)
+
+proc toInt*(x: Rational): int =
+  ## Convert a rational number `x` to an int. Conversion rounds if `x` does not
+  ## contain an integer value.
+  round(toFloat(x))
```
Comment (ReneSac) [inline lib/pure/rational.nim]: Why not simply integer division? Converting to float can mean reduced precision for things like 9_010_199_254_740_992 / 1. And this conversion back and forth to float probably makes it a bit slower too (not really relevant).

The only upside is better rounding, I guess. But you can get it too using the remainder. Maybe add an optional argument for the user define the round mode.

File: lib/pure/rational.nim [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+  if x.den > 0:
+    x.num = x.num div common
+    x.den = x.den div common
+  elif x.den < 0:
+    x.num = -x.num div common
+    x.den = -x.den div common
+  else:
+    raise newException(DivByZeroError, "division by zero")
+
+proc `+` *(x, y: Rational): Rational =
```
Comment (ReneSac) [inline lib/pure/rational.nim]: You can define `+` (and `*`, etc) using the `+=` and avoid all that code duplication:

``` nimrod
proc `+` *(x: Rational, y: Rational | int): Rational =
  result = x
  result += y

proc `+` * (x: int, y: Rational): Rational =
  result = y
  result += x
```

File: lib/pure/rational.nim [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  let common = lcm(x.den, y.den)
+  result.num = common div x.den * x.num + common div y.den * y.num
+  result.den = common
+  reduce(result)
+
+proc `+` *(x: Rational, y: int): Rational =
+  ## Add rational `x` to int `y`.
+  result.num = x.num + y * x.den
+  result.den = x.den
+
```
Comment (ReneSac) [inline lib/pure/rational.nim]: You can define `+` using the `+=` and avoid all that code duplication:

``` nimrod
proc `+` *(x: Rational, y: Rational | int): Rational =
  result = x
  result += y

proc `+` *(x: int, y: Rational): Rational =
  result = y
  result += x
  # or result = y + x (but one more level of function calls if not inlined)
```
Similarly for  `*`, etc.

**Comment by def-** [comment]
@ReneSac: I mainly oriented on the complex module, which is why I explicitly stated `+` and other procs. You're right about calling it rationals instead I guess.

**Comment by ReneSac** [comment]
Someone doing it wrong don't justify you too doing it. There is no excuse for ctrl-c + ctrl+v coding. If it is because of performance, then you should use {.inline.}, or templates, etc.

File: lib/pure/rationals.nim [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+  result.num = x.num
+  result.den = x.den * y
+  reduce(result)
+
+proc `/`*(x: int, y: Rational): Rational =
+  ## Divide int `x` by Rational `y`.
+  result.num = y.num
+  result.den = x * y.den
+  reduce(result)
+
```
Comment (ReneSac) [inline lib/pure/rationals.nim]: You need more tests:
``` Nimrod
when isMainModule:
  var i = 3
  var a = (1, 2)
  assert (i/a == (6 , 1) )
```
Fails.

File: lib/pure/rationals.nim [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  result.den = 1
+
+proc toFloat*(x: Rational): float =
+  ## Convert a rational number `x` to a float.
+  x.num / x.den
+
+proc toInt*(x: Rational): int =
+  ## Convert a rational number `x` to an int. Conversion rounds down if `x`
+  ## does not contain an integer value.
+  x.num div x.den
```
Comment (ReneSac) [inline lib/pure/rationals.nim]: It rounds towards zero, not really "down".

``` Nimrod
assert( toInt( (5,3) ) == 1 )
assert( toInt( (-5,3) ) == -1 )
```

**Comment by BlaXpirit** [comment]
Please don't add `shuffle` to `math`. We should move in the opposite direction: the `random` functions that are present in it should be deprecated, because they're [harmful](https://github.com/Araq/Nim/issues/1506).

**Comment by def-** [comment]
@BlaXpirit: Thanks, didn't know about these issues.

**Comment by ReneSac** [comment]
Two points from the irc discussion: maybe you should use `distinct tuple` for the rational type, and the additional random procs probably shouldn't be on a math lib (nor the original random, but anyway).

**Comment by def-** [comment]
@ReneSac: Looks like the old tuple behaviour is back, so no distinct necessary in my opinion: https://github.com/Araq/Nim/commit/abb738146aea1b9692362bf8048ca23a93bd9fa5

Yes, random was supposed to be visible before.

File: lib/pure/math.nim [hunk 6]
```diff
@@ -329,6 +329,30 @@ proc standardDeviation*(s: RunningStat): float =
...
+  while y != 0:
+    if (y and 1) != 0:
+      result *= x
+    y = y shr 1
+    x *= x
+
+proc gcd*[T](x, y: T): T =
+  ## Computes the greatest common divisor of ``x`` and ``y``.
+  if y != 0:
+    gcd(y, x mod y)
```
Comment (Araq) [inline lib/pure/math.nim]: This is not a tail recursion. Use iteration instead please.

Comment (def-) [inline lib/pure/math.nim]: Alright, done.

File: doc/lib.txt [hunk 7]
```diff
@@ -185,6 +185,9 @@ Math libraries
 * `complex <complex.html>`_
   This module implements complex numbers and their mathematical operations.
 
+* `rational <rational.html>`_
```
Comment (Araq) [inline doc/lib.txt]: well the module is 'rationals' now.

**Comment by oderwat** [comment]
May I ask why this has stalled? I would love to use rationals in a project of my company!

File: lib/pure/rationals.nim [hunk 8]
```diff
@@ -0,0 +1,255 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2015 Andreas Rumpf
```
Comment (Araq) [inline lib/pure/rationals.nim]: I'm not the author of this!

File: lib/pure/rationals.nim [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+  result.den = common
+  reduce(result)
+
+proc `+` *(x: Rational, y: int): Rational =
+  ## Add rational `x` to int `y`.
+  result.num = x.num + y * x.den
+  result.den = x.den
+
+proc `+` *(x: int, y: Rational): Rational =
+  ## Add int `x` to tational `y`.
```
Comment (dom96) [inline lib/pure/rationals.nim]: typo here

**Comment by ReneSac** [comment]
Question: would it be too difficult to make rationals more generic? And should it be done now? For example: suppose I want a minimal precision of a 64bit integer, even on 32bit systems (working with tick counts from system clocks).

**Comment by def-** [comment]
I tried to make it generic, you can see it here: https://github.com/def-/Nim/tree/extend-math-2

Seems to work from an initial look, would still need more work though. In that case complex numbers should be generic too.

**Comment by def-** [comment]
Rationals being a tuple causes problems, see https://github.com/Araq/Nim/issues/2178. Should it be a distinct tuple or object instead?

**Comment by Araq** [comment]
An object of course. A "distinct tuple" is almost an oxymoron.

**Pull request merged**